### PR TITLE
Transducer float16 loss bug fix

### PR DESCRIPTION
**Pull request by GNroy**
I've just encountered a problem with transducer training in the float16 mode. It seems like warprnnt_pytorch doesn't support float16.

This fix casts a tensor of predicted sequences to float32 if necessary.

Thanks,
Aleksandr Laptev.

### CODE REVIEW

**Review by b-flo — COMMENTED** [review]
Thanks for the fix! I'm wondering some things though:
  - I'm not sure what I prefer: casting or raising an error. We should at least warns users I think. @sw005320 do you have an opinion on that?
  - What about others types? To be honest I never tested these options.
  - #2444 introduces `warp-rnnt`, could you also test it with the same options?

If you don't have time for that we can merge this PR and I will make the appropriate modifications on my side, it will conflict with #2444 anyways. Otherwise you will have to wait a bit after we merge the other, test and last modifications are almost done.

**Comment by sw005320** [comment]
> * I'm not sure what I prefer: casting or raising an error. We should at least warns users I think. @sw005320 do you have an opinion on that?

If casting makes it work, I think it's better to do it. Adding a warning is a nice idea.

**Comment by ShigekiKarita** [comment]
I don't think the warning is needed because our ctc and pytorch/nvidia amp also cast implicitly
https://github.com/espnet/espnet/blob/c0d6029ccd3230d84f1d0f8d3413382304238475/espnet/nets/pytorch_backend/ctc.py#L106-L109
https://github.com/NVIDIA/apex/blob/93cabd5df060b3783228dd75941ad359be5d391d/apex/amp/lists/functional_overrides.py#L67

**Comment by ShigekiKarita** [comment]
> What about others types? To be honest I never tested these options.

For warpctc, we already tested here https://github.com/espnet/espnet/blob/6f6452a3891d11b0b0306084dd46a464ddf8e25c/test/test_train_dtype.py#L69

**Review by ShigekiKarita — APPROVED** [review]

**Comment by b-flo** [comment]
Thanks for the feedback @ShigekiKarita !

> I don't think the warning is needed because our ctc and pytorch/nvidia amp also cast implicitly

In general I prefer to warn the user when an option he requested for is not working but fine with me! 

> What about others types? To be honest I never tested these options.
> 
> > For warpctc, we already tested here 

I was wondering for `warp-transducer` as we don't test that option currently

**Comment by GNroy** [comment]
Thanks for the feedback!
@b-flo @ShigekiKarita I think this PR can wait until #2444 is merged. After that, I can also test warp-rnnt for being able to handle float16.
Until then I'll try to add tests for transducer to test/test_train_dtype.py

File: tools/installers/install_warp-rnnt.sh
Comment (b-flo) [inline tools/installers/install_warp-rnnt.sh:13]:
```suggestion
if ! [[ "$cuda_ver" =~ ^(10.0|10.1|10.2)$ ]]; then
```

I tested all cuda versions, I'm surprised you could install repo with `CUDA_VERSION=10.1` or `10.2`?

File: tools/installers/install_warp-rnnt.sh
Comment (b-flo) [inline tools/installers/install_warp-rnnt.sh:18]: I would prefer not relying on pypi package. If needed we can quickly fix through PR or our own fork and not wait for pypi package new version.

File: tools/installers/install_warp-rnnt.sh
Comment (GNroy) [reply tools/installers/install_warp-rnnt.sh:13]: I can't say I thoroughly tested it, but the warp-rnnt unit test worked. Though I installed warp-rnnt with pip, if it matters.

File: tools/installers/install_warp-rnnt.sh
Comment (GNroy) [reply tools/installers/install_warp-rnnt.sh:18]: Ok, I'll revert the original installation back

File: tools/installers/install_warp-rnnt.sh
Comment (GNroy) [reply tools/installers/install_warp-rnnt.sh:13]: No, it is `python -m warp_rnnt.test`

Comment (b-flo) [reply tools/installers/install_warp-rnnt.sh:13]: Sorry I read too quickly!
Hum, every install (from repo or pypi) with a cuda version != 10.0 ended-up with an ImportError when I tested. Could you verify? If not, I'll do it tomorrow

Comment (GNroy) [reply tools/installers/install_warp-rnnt.sh:13]: Welp, I have no ImportError in cuda 10.1 environment. I'm going to try this also with a clean miniconda environment to verify though.

Comment (b-flo) [reply tools/installers/install_warp-rnnt.sh:13]: Sorry, I edited my message `ImportError` -> `error`! The error I encountered was:

```
  File "/home/fboye/espnet/tools/venv/lib/python3.7/site-packages/warp_rnnt-0.4.0-py3.7-linux-x86_64.egg/warp_rnnt/__init__.py", line 2, in <module>
    import warp_rnnt._C as core
ImportError: /home/fboye/espnet/tools/venv/lib/python3.7/site-packages/warp_rnnt-0.4.0-py3.7-linux-x86_64.egg/warp_rnnt/_C.cpython-37m-x86_64-linux-gnu.so: undefined symbol: cudaSetupArgument
```

Comment (GNroy) [reply tools/installers/install_warp-rnnt.sh:13]: Just checked with clean cuda 10.2 torch 1.6 env. I am using Ubuntu 18.04.
```
import warp_rnnt._C as core
...
costs, grads = core.rnnt_loss(
...             xs.cuda(), ys.cuda(),
...             xn.cuda(), yn.cuda())

>>> costs
tensor([4.2742], device='cuda:0')
>>> grads
tensor([[[[ 0., -1.,  0.,  0.,  0.],
          [ 0.,  0., -1.,  0.,  0.],
          [-1.,  0.,  0.,  0.,  0.]]]], device='cuda:0')```

Comment (b-flo) [reply tools/installers/install_warp-rnnt.sh:13]: Could you also try with torch==1.4 please? If it works, let's keep that condition (with the fix I proposed or change `||` to `&&`)!
I may have made a mistake when testing, good to know if it also works for other cases!

Thanks!

Comment (GNroy) [reply tools/installers/install_warp-rnnt.sh]: Sorry, I pushed the fix just before reading your last comment.
torch 1.4 was my initial setup, where everything worked. I just used 1.6 for a wider check.

Comment (b-flo) [reply tools/installers/install_warp-rnnt.sh]: OK, thank you!

**Comment by GNroy** [comment]
@sw005320 @ShigekiKarita I suppose this PR is ready to be merged.

**Review by sw005320 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
